In [1]:
# !pip install qibo

## Quantum Fourier Transform to test installation

In [2]:
from qibo.models import QFT

# Create a QFT circuit with 15 qubits
circuit = QFT(15)

# Simulate final state wavefunction default initial state is |00>
final_state = circuit()

[Qibo 0.2.1|INFO|2023-10-24 21:50:01]: Using tensorflow backend on /device:CPU:0


In [3]:
print(final_state)


(0.00552+0j)|000000000000000> + (0.00552+0j)|000000000000001> + (0.00552+0j)|000000000000010> + (0.00552+0j)|000000000000011> + (0.00552+0j)|000000000000100> + (0.00552+0j)|000000000000101> + (0.00552+0j)|000000000000110> + (0.00552+0j)|000000000000111> + (0.00552+0j)|000000000001000> + (0.00552+0j)|000000000001001> + (0.00552+0j)|000000000001010> + (0.00552+0j)|000000000001011> + (0.00552+0j)|000000000001100> + (0.00552+0j)|000000000001101> + (0.00552+0j)|000000000001110> + (0.00552+0j)|000000000001111> + (0.00552+0j)|000000000010000> + (0.00552+0j)|000000000010001> + (0.00552+0j)|000000000010010> + (0.00552+0j)|000000000010011> + ...


In [4]:
# Create a QFT circuit with 2 qubits
circuit = QFT(2)
final_state = circuit()
print()
print(final_state)


(0.5+0j)|00> + (0.5+0j)|01> + (0.5+0j)|10> + (0.5+0j)|11>


# Add two Hadamard gates:

Here is an example of a circuit with 2 qubits:

Starting with a uniform (máximum) superposition of all possible states for two qubits:

$|\psi\rangle = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$

Each of the base states (|00⟩, |01⟩, |10⟩, and |11⟩) has an equal probability of being measured. In python this state is:

```python
initial_state = np.ones(4) / np.sqrt(4)
```

**Circuit:**

|+> ----- H ----- |0>

|+> ----- H ----- |0>

In [7]:
import numpy as np
from qibo import Circuit, gates

# Construct the circuit
c = Circuit(2)
# Add some gates
c.add(gates.H(0))
c.add(gates.H(1))

# Define an initial state (optional - default initial state is |00>)
initial_state = np.ones(4) / 2.0

# Execute the circuit and obtain the final state
result = c(initial_state) # c.execute(initial_state) also works

# Prints the final state in Dirac form:
print('Final state:', result)
print()


# Prints the final state in matrix form:
print(result.state()) # should print `tf.Tensor([1, 0, 0, 0])`
print(result.state(numpy=True)) # should print `np.array([1, 0, 0, 0])`

Final state: (1+0j)|00>

tf.Tensor([1.+0.j 0.+0.j 0.+0.j 0.+0.j], shape=(4,), dtype=complex128)
[1.+0.j 0.+0.j 0.+0.j 0.+0.j]


## Freeze circuit and querry for different initial states

Freeze the circuit and just query for different initial states then you can use the `Circuit.compile()` method which will improve evaluation performance

|0> ----- X ----- |1>

|0> ----- X ----- |1>

In [6]:
# !pip install qibo tensorflow

# !pip uninstall qibo tensorflow

In [4]:
import numpy as np
from qibo import Circuit, gates

In [18]:
import qibo

# This can be beneficial if you're going to utilize TensorFlow's capabilities, such as automatic differentiation, in your quantum simulations.
qibo.set_backend("tensorflow")

c = Circuit(2)
c.add(gates.X(0))
c.add(gates.X(1))
c.add(gates.CU1(0, 1, 0.1234))
c.compile()

# For a specific initial state:
init_state = np.ones(4) / 2.0 + 1
norm = np.linalg.norm(init_state)
print(c(init_state), '\nNorm=',norm)

[Qibo 0.2.1|INFO|2023-10-22 15:03:18]: Using tensorflow backend on /device:CPU:0


(1.5+0j)|00> + (1.5+0j)|01> + (1.5+0j)|10> + (1.48859+0.18463j)|11> 
Norm= 3.0


For the first state:

$|\psi_i\rangle = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$

$CU1|X\otimes X|\psi_i\rangle = \frac{1}{2}(\exp[\theta]|11\rangle + |10\rangle + |01\rangle + |00\rangle)$

with $\theta = 0.1234$

Let's check the output fot the state $|11\rangle$:

In [26]:
th=0.1234

3/2*(np.cos(th) + np.sin(th)*1j)

(1.4885938150573366+0.1846305873170644j)

For 100 different initial states:

q0 ---X-------o---

               |

q1 ---X-------•---

In [27]:
c = Circuit(2)
c.add(gates.X(0))
c.add(gates.X(1))
c.add(gates.CU1(0, 1, 0.1234))
c.compile()

for i in range(100):
    init_state = np.ones(4) / 2.0 + i
    norm = np.linalg.norm(init_state)
    print('i=',i,':',c(init_state),'\n  Norm=', norm)

i= 0 : (0.5+0j)|00> + (0.5+0j)|01> + (0.5+0j)|10> + (0.4962+0.06154j)|11> 
  Norm= 1.0
i= 1 : (1.5+0j)|00> + (1.5+0j)|01> + (1.5+0j)|10> + (1.48859+0.18463j)|11> 
  Norm= 3.0
i= 2 : (2.5+0j)|00> + (2.5+0j)|01> + (2.5+0j)|10> + (2.48099+0.30772j)|11> 
  Norm= 5.0
i= 3 : (3.5+0j)|00> + (3.5+0j)|01> + (3.5+0j)|10> + (3.47339+0.4308j)|11> 
  Norm= 7.0
i= 4 : (4.5+0j)|00> + (4.5+0j)|01> + (4.5+0j)|10> + (4.46578+0.55389j)|11> 
  Norm= 9.0
i= 5 : (5.5+0j)|00> + (5.5+0j)|01> + (5.5+0j)|10> + (5.45818+0.67698j)|11> 
  Norm= 11.0
i= 6 : (6.5+0j)|00> + (6.5+0j)|01> + (6.5+0j)|10> + (6.45057+0.80007j)|11> 
  Norm= 13.0
i= 7 : (7.5+0j)|00> + (7.5+0j)|01> + (7.5+0j)|10> + (7.44297+0.92315j)|11> 
  Norm= 15.0
i= 8 : (8.5+0j)|00> + (8.5+0j)|01> + (8.5+0j)|10> + (8.43536+1.04624j)|11> 
  Norm= 17.0
i= 9 : (9.5+0j)|00> + (9.5+0j)|01> + (9.5+0j)|10> + (9.42776+1.16933j)|11> 
  Norm= 19.0
i= 10 : (10.5+0j)|00> + (10.5+0j)|01> + (10.5+0j)|10> + (10.42016+1.29241j)|11> 
  Norm= 21.0
i= 11 : (11.5+0j)|00> +

## Circuir Summary


q0 ---H-------------o------------------------•----------

                    |                 |

q1 ---H--------------|-------o----------------•-------

                    |     |           |

q2 ------------------X-------X-------H-------X-------

In [28]:
c = Circuit(3)
c.add(gates.H(0))
c.add(gates.H(1))
c.add(gates.CNOT(0, 2))
c.add(gates.CNOT(1, 2))
c.add(gates.H(2))
c.add(gates.TOFFOLI(0, 1, 2))
print(c.summary())

Circuit depth = 5
Total number of gates = 6
Number of qubits = 3
Most common gates:
h: 3
cx: 2
ccx: 1


Why `Circuit depth = 5`?

**Try to solve this!**

In [29]:
common_gates = c.gate_names.most_common()
common_gates

[('h', 3), ('cx', 2), ('ccx', 1)]

In [30]:
most_common_gate = common_gates[0][0]
most_common_gate

'h'

In [31]:
all_h_gates = c.gates_of_type(gates.H)
all_h_gates

[(0, <qibo.gates.gates.H at 0x1d9627ee410>),
 (1, <qibo.gates.gates.H at 0x1d962766010>),
 (4, <qibo.gates.gates.H at 0x1d96276e690>)]

## Meassurements

Add 2 qubits in |0>

**Circuit:**

$|0\rangle$ ----- X ----- $|1\rangle$

$|0\rangle$ ------------- $|0\rangle$

In [1]:
import numpy as np
from qibo import Circuit, gates

[Qibo 0.2.1|INFO|2023-10-22 17:51:39]: Using tensorflow backend on /device:CPU:0


In [41]:
c = Circuit(2)
c.add(gates.X(0))

# Add a measurement register on both qubits
c.add(gates.M(0, 1))

# Execute the circuit with the default initial state |00>.
result = c(nshots=100)

print(result)

(1+0j)|10>


In [3]:
result.samples(binary=True)

array([[1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1,

**In binary form:**

Given a binary number with digits $( b_n, b_{n-1}, \ldots, b_2, b_1, b_0 )$, with $b=0$ or $1$ the conversion to its decimal representation $D$ is:

$ D = b_0 \times 2^0 + b_1 \times 2^1 + b_2 \times 2^2 + \ldots + b_n \times 2^n $

<br>

For example, for the binary number $ 1011 $:

$ D = 1 \times 2^0 + 1 \times 2^1 + 0 \times 2^2 + 1 \times 2^3 $
$ = 1 + 2 + 0 + 8 = 11 $

<br>

For the state $|10\rangle$, $b_1=1$ and $b_0=0$:

$ D = 0 \times 2^0 + 1 \times 2^1 = 2$

In [4]:
result.samples(binary=False)

<tf.Tensor: shape=(100,), dtype=int32, numpy=
array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])>

In [5]:
print(result.frequencies(binary=True))
print(result.frequencies(binary=False))

Counter({'10': 100})
Counter({2: 100})


### Collect measurement results grouped according to registers

**Circuit:**

$|0\rangle ------ X ------ |1\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------X------  |1\rangle$

In [46]:
c = Circuit(5)
c.add(gates.X(0))
c.add(gates.X(4))
c.add(gates.M(0, 1, register_name="A"))
c.add(gates.M(3, 4, register_name="B"))
result = c(nshots=100)

print(result)

(1+0j)|10001>


Register **A** meassures $|10\rangle$,

$\qquad D=2$ (100-times)

Register **B** meassures $|01\rangle$,

$\qquad D=b_0*2^0+b_1*2^1=1+0*2=1$ (100-times)

In [7]:
result.samples(binary=False, registers=True)

{'A': <tf.Tensor: shape=(100,), dtype=int32, numpy=
 array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])>,
 'B': <tf.Tensor: shape=(100,), dtype=int32, numpy=
 array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])>}

Not in binary:

In [8]:
result.frequencies(binary=True, registers=True)

{'A': Counter({'10': 100}), 'B': Counter({'01': 100})}

Setting `registers=False` (default option) will ignore the registers 

In [9]:
# Both prints the same
print(result.frequencies(binary=True, registers=False))
print(result.frequencies(binary=True))

Counter({'1001': 100})
Counter({'1001': 100})


### Define registers of multiple qubits

**Circuit:**

$|0\rangle ------ X ------ |1\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------------- |0\rangle$

$|0\rangle ------X------  |1\rangle$

In [51]:
c = Circuit(5)
c.add(gates.X(0))
c.add(gates.X(4))
c.add(gates.M(*[0,1,2,4], register_name="A"))
result = c(nshots=100)

# Draw the circuit:
print(c.draw())
print()

# Prints final state (Dirac form)
print(result)
print()

# Prints final state (matrix form)
print(result.state())

q0: ─X─M─
q1: ───M─
q2: ───M─
q3: ───|─
q4: ─X─M─

(1+0j)|10001>

tf.Tensor(
[0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j
 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j
 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j
 0.+0.j 0.+0.j], shape=(32,), dtype=complex128)


The final state vector is still accessible via `qibo.states.CircuitResult.state()`. 
Note that the state vector accessed this way corresponds to the state as if no measurements occurred, that is the state is not collapsed during the measurement.
This is because **measurement gates are only used to sample bitstrings** and do not have any effect on the state vector.

For applications that require the state vector to be collapsed during measurements, Below we will do it!

<br>

**Unmeasured qubits are ignored by the measurement objects.**

$D(|q_0, q_1, q_2, q_4\rangle=|1001\rangle) = 1 + 1*2^3 = 9$

In [30]:
print(result.frequencies(binary=True))
print(result.frequencies(binary=False))

Counter({'1001': <tf.Tensor: shape=(), dtype=int64, numpy=100>})
Counter({9: <tf.Tensor: shape=(), dtype=int64, numpy=100>})


**Also, the order that qubits appear in the results is defined by the order the user added the measurements and not the qubit ids.**

In [48]:
import qibo

print(qibo.get_metropolis_threshold()) # prints 100000
qibo.set_metropolis_threshold(int(1e8))
print(qibo.get_metropolis_threshold()) # prints 10^8

100000
100000000


In [55]:
from qibo.models import QFT

c = QFT(2)
print(c.draw())
print()
print(c())

q0: ─H─U1───x─
q1: ───o──H─x─

(0.5+0j)|00> + (0.5+0j)|01> + (0.5+0j)|10> + (0.5+0j)|11>


In [ ]:
c = Circuit(2)
c.add(gates.H(0))
c.add(gates.X(4))
c.add(gates.M(*[0,1,2,4], register_name="A"))
result = c(nshots=100)

# Draw the circuit:
print(c.draw())
print()

# Prints final state (Dirac form)
print(result)
print()

# Prints final state (matrix form)
print(result.state())

For simulation precision see the final part:

https://qibo.science/qibo/stable/code-examples/examples.html

In [56]:
c = QFT(5)
print(c.draw())

q0: ─H─U1─U1─U1─U1───────────────────────────x───
q1: ───o──|──|──|──H─U1─U1─U1────────────────|─x─
q2: ──────o──|──|────o──|──|──H─U1─U1────────|─|─
q3: ─────────o──|───────o──|────o──|──H─U1───|─x─
q4: ────────────o──────────o───────o────o──H─x───


    '''
    q0: ─H─U1─U1─U1─U1───────────────────────────x───
    q1: ───o──|──|──|──H─U1─U1─U1────────────────|─x─
    q2: ──────o──|──|────o──|──|──H─U1─U1────────|─|─
    q3: ─────────o──|───────o──|────o──|──H─U1───|─x─
    q4: ────────────o──────────o───────o────o──H─x───
    '''